### ISSUE TITLE
`torch.zeros` hangs on TPU

### ISSUE THREAD

**Issue by zcain117** [PyTorchLightning/pytorch-lightning#8183]
## 🐛 Bug

PyTorch Lightning hangs when calling `torch.zeros` of a sufficiently large size, given that the input data is sufficiently large

* Hangs on both TPUVM and 2VM. In my earlier experiments, I anecdotally noticed that 2VM could go slightly higher on data size (compared to TPUVM) before hanging started, but right now they both hang given the repro script below.
* Hangs only when input data is sufficiently large. For example in the attached script, I used input data of shape [6, 3, 256, 256] and am able to repro the hang every time. Previously I tried a smaller input of shape [1, 784] and the same size of `torch.zeros` and I don't see the hang
* Hangs only when `torch.zeros` call is sufficiently large. I tried `torch.zeros([256, 81])` and it does not hang but `torch.zeros([256, 8192])` hangs

## Repro script

Here is a simple repro script. It's almost identical to the `LitAutoEncoder` except:
1. uses fake data generator to make a large input data shape
2. calls `torch.zeros` during the training loop

```
import os
import torch
from torch import nn
import torch.nn.functional as F
from torchvision import transforms
from torchvision.datasets import MNIST
from torch.utils.data import DataLoader, random_split
import pytorch_lightning as pl

import torch_xla.utils.utils as xu
import torch_xla.core.xla_model as xm

class LitAutoEncoder(pl.LightningModule):

    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(
            nn.Linear(28*28, 64),
            nn.ReLU(),
            nn.Linear(64, 3)
        )
        self.decoder = nn.Sequential(
            nn.Linear(3, 64),
            nn.ReLU(),
            nn.Linear(64, 28*28)
        )

    def forward(self, x):
        # in lightning, forward defines the prediction/inference actions
        embedding = self.encoder(x)
        return embedding
    def training_step(self, batch, batch_idx):
        # training_step defined the train loop.
        # It is independent of forward
        x, y = batch
        myzeros = torch.zeros([256, 8192]).to(x)
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = F.mse_loss(x_hat, x)
        # Logging to TensorBoard by default
        self.log('train_loss', loss)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer

batch_size=6
img_size=256
train_loader = xu.SampleGenerator(
    data=(torch.zeros(batch_size, 3, img_size , img_size ),
    torch.zeros(batch_size, dtype=torch.int64)),
    sample_count=1200000 // batch_size // xm.xrt_world_size())
val_loader = xu.SampleGenerator(
    data=(torch.zeros(batch_size, 3, img_size , img_size ),
    torch.zeros(batch_size, dtype=torch.int64)),
    sample_count=50000 // batch_size // xm.xrt_world_size())                           
 

# init model
autoencoder = LitAutoEncoder()

# most basic trainer, uses good defaults (auto-tensorboard, checkpoints, logs, and more)
# trainer = pl.Trainer(gpus=8) (if you have GPUs)
trainer = pl.Trainer(tpu_cores=8)
trainer.fit(autoencoder, train_loader)
```

### To Reproduce
TPUVM setup:
1. Create TPUVM: `gcloud alpha compute tpus tpu-vm create $TPUVM_NAME \
  --zone=europe-west4-a \
  --accelerator-type=v3-8 \
  --version=v2-nightly --project=$MY-PROJECT`
2. `gcloud alpha compute tpus tpu-vm ssh $TPUVM_NAME --zone europe-west4-a --project=$MY_PROJECT`
3. ctrl-f the contents of the repro script in a file on the TPUVM (assumed `repro.py`)
4. (on the TPUVM terminal) `git clone https://github.com/PyTorchLightning/pytorch-lightning.git`
5. (on the TPUVM terminal) `cd pytorch-lightning && pip3 install .`
6. (on the TPUVM terminal) `export XRT_TPU_CONFIG="localservice;0;localhost:51011"`
7. (optional if you want to see output from the TPU. This is helpful so you can tell whether the TPU is working on something or if it's the python side of things that's hanging) `export TF_CPP_VMODULE=tensor=5,computation_client=5,xrt_computation_client=5,aten_xla_type=1 && export TF_CPP_MIN_LOG_LEVEL=0`
8. (on the TPUVM terminal) `python3 repro.py`

2VM setup:
1. Create TPU, use `pytorch-1.8` as the software type
2. Create VM, use same zone as TPU. For the Boot Disk, choose "Deep Learning on Linux" for operating system and choose "Debian GNU/Linux 9 Stretch + PyTorch/XLA" for version
3. SSH into VM
4. (on VM) `conda activate torch-xla-1.8.1`
5. (on VM) `git clone https://github.com/PyTorchLightning/pytorch-lightning.git`)
6. (on VM) `cd pytorch-lightning/ && pip install .`
7. copy contents of repro script into e.g. `repro.py` on the VM
8.  (on VM) `export TPU_IP_ADDRESS=xx.xx.xx` (find this on the TPU you created on the TPU Google Cloud UI)
9. (on VM) `export XRT_TPU_CONFIG="tpu_worker;0;$TPU_IP_ADDRESS:8470"`
10. (optional if you want to see output from the TPU. This is helpful so you can tell whether the TPU is working on something or if it's the python side of things that's hanging) `export TF_CPP_VMODULE=tensor=5,computation_client=5,xrt_computation_client=5,aten_xla_type=1 && export TF_CPP_MIN_LOG_LEVEL=0`
11. (on VM) `python repro.py`

**Comment by kaushikb11** [comment]
@zcain117 Thanks for reporting the Issue! Did you face this issue with Pytorch XLA as well?

**Comment by zcain117** [comment]
No I don't see it when using PyTorch XLA directly.

For example, I opened our "Getting Started with PyTorch on Cloud TPUs" colab and ran this cell:

```
dev = xm.xla_device()
t1 = torch.rand([6, 3, 256, 256], device = dev)
t2 = torch.zeros([256, 8192]).to(t1)
print(t1)
print(t2)
```

it successfully prints `t1` and `t2` and both show `device='xla:1'`

Colab machines are quite small too in terms of RAM. I think they are using v2-8 TPU which has smaller HBM memory available as well

**Comment by kaushikb11** [comment]
@zcain117 I tried running it on a single-core, faced this issue
```
Epoch 0:   0%|                                                                                                                                                                                     | 0/200000 [00:00<?, ?it/s]https://symbolize.stripped_domain/r/?trace=8cb71200,7fbc7887820f,1&map=
*** SIGSEGV (@0x8cb71200), see gl__________25#s15 received by PID 24736 (TID 24736) on cpu 25; stack trace: ***
PC: @         0x8cb71200  (unknown)  (unknown)
    @     0x7fb9f1f0a1e0        976  (unknown)
    @     0x7fbc78878210  2089952272  (unknown)
    @                0x2  (unknown)  (unknown)
https://symbolize.stripped_domain/r/?trace=8cb71200,7fb9f1f0a1df,7fbc7887820f,1&map=ca1b7ab241ee28147b3d590cadb5dc1b:7fb9e520b000-7fb9f223db20
E0629 14:24:22.031031   24736 coredump_hook.cc:292] RAW: Remote crash data gathering hook invoked.
E0629 14:24:22.031047   24736 coredump_hook.cc:384] RAW: Skipping coredump since rlimit was 0 at process start.
E0629 14:24:22.031055   24736 client.cc:222] RAW: Coroner client retries enabled (b/136286901), will retry for up to 30 sec.
E0629 14:24:22.031062   24736 coredump_hook.cc:447] RAW: Sending fingerprint to remote end.
E0629 14:24:22.031069   24736 coredump_socket.cc:124] RAW: Stat failed errno=2 on socket /var/google/services/logmanagerd/remote_coredump.socket
E0629 14:24:22.031077   24736 coredump_hook.cc:451] RAW: Cannot send fingerprint to Coroner: [NOT_FOUND] Missing crash reporting socket. Is the listener running?
E0629 14:24:22.031080   24736 coredump_hook.cc:525] RAW: Discarding core.
E0629 14:24:22.251441   24736 process_state.cc:771] RAW: Raising signal 11 with default behavior
Segmentation fault (core dumped)
```

**Comment by zcain117** [comment]
Regarding my previous comment about Colab: I realized that cell I shared was probably using just 1 core. So I tried editing https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py so that it creates `torch.zeros([256, 8192], device=device)` during training. It is able to run with no issue. So I think this is not a problem for plain pytorch/xla even in 8-core case.

Regarding your question: yes the `repro.py` will segfault if it makes it past the `torch.zeros` calls and continues the forward pass. This is because the input data shape does not match the `self.encoder` as defined in the model. I didn't both changing the  encoder when I copied [this example](https://pytorch-lightning.readthedocs.io/en/stable/starter/new-project.html) since it wasn't relevant to the hanging part. If you wanted training to run fully then you could change the encoder's input size `nn.Linear(28*28, 64)` to be some size that works and maybe change the `x = x.view(x.size(0), -1)` in coordination with the encoder change

**Comment by zcain117** [comment]
ah actually you'd also need to change the decoder so that it produces a final output `x_hat` that's the same shape as `x`. Otherwise the loss computation will probably crash: `loss = F.mse_loss(x_hat, x)`

**Comment by zcain117** [comment]
so you could use this for the encoder/decoder if you want training to proceed:

```
class LitAutoEncoder(pl.LightningModule):

    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(
            nn.Linear(3*256*256, 64),
            nn.ReLU(),
            nn.Linear(64, 3)
        )
        self.decoder = nn.Sequential(
            nn.Linear(3, 64),
            nn.ReLU(),
            nn.Linear(64, 3*256*256)
        )
```

just tried on single core (still including the large `torch.zeros`) and it's able to train either with `batch_size=6` or with `batch_size=1`. 8-core still hangs at the `torch.zeros` call, even when using `batch_size=1`

**Comment by kaushikb11** [comment]
@zcain117 Thanks for the update! I tested it, and it does work for single TPU core, but I do get this message at the end of the training (batch_size=6)
```
tcmalloc: large alloc 1179049984 bytes == 0xa1826000 @  0x7f9026cee680 0x7f9026d0eff4 0x7f9026805309 0x7f9026806222 0x7f90268062d8 0x7f9026806700 0x7f8dae096dbe 0x7f8da3906db7 0x7f8da38dc25f 0x7f8da38eb5a5 0x7f8da39074e4 0x7f8da3907004 0x7f8da390787f 0x7f8da38dbe0c 0x7f8da38e148b 0x7f8da38e24ee 0x7f8da38c5df6 0x7f90267edd84 0x7f9026ac4609 0x7f9026c00293
```

**Comment by zcain117** [comment]
right, that's just a warning from `tcmalloc`. I tried both `tcmalloc` (which is preinstalled on the TPUVM) and a different malloc implementation (which you can try by running `unset LD_PRELOAD`) and both result in hanging

**Comment by kaushikb11** [comment]
@zcain117 You would need to call `myzeros = torch.zeros([256, 8192], device=self.device)` inside the methods. It would work as expected!

**Comment by zcain117** [comment]
Thanks, that works!

I previously tried `torch.zeros([256, 8192], device=x.device)` where `x` was another tensor that was already on TPU. That version I still couldn't get working.

But `device=self.device` seems to work, so that's all I need
